src/bt.rs

```diff
@@ -0,0 +1,426 @@
+//!
+//! ```txt
+//!               ()=@
+//!          /              \
+//!       (0)=a           (1)=b
+//!     /      \         /      \
+//! (00)=c  (01)=d   (10)=e  (11)=f
+//! ```
+//!
+//! ```
+//! use prio::bt::BinaryTree;
+//! use bitvec::{bits,prelude::Lsb0};
+//! let mut tree = BinaryTree::new('@');
+//! tree.insert(bits!(0), 'a');
+//! tree.insert(bits!(1), 'b');
+//! tree.insert(bits!(0, 0), 'c');
+//! tree.insert(bits!(0, 1), 'd');
+//! tree.insert(bits!(1, 0), 'e');
+//! tree.insert(bits!(1, 1), 'f');
+//! ```
+
+use core::fmt::Debug;
+use core::fmt::{Display, Formatter};
+use std::io::Cursor;
+
+use bitvec::slice::BitSlice;
+
+use crate::codec::{CodecError, Decode, Encode};
+
+/// Used to indicate a traversal path on the binary tree.
+pub type Path = BitSlice;
+
+/// Represents an append-only binary tree.
+pub struct BinaryTree<V> {
+    root: Node<V>,
+}
+
+impl<V> BinaryTree<V> {
+    /// Creates a binary tree initialized with a value as the root node.
+    pub fn new(value: V) -> Self {
+        Self {
+            root: Node::new(value),
+        }
+    }
+
+    /// Inserts the value in the first available node along the path.
+    /// No insertion occurs if the nodes along the path are already filled.
+    pub fn insert(&mut self, path: &Path, value: V) {
+        self.root.insert(path, value)
+    }
+
+    /// Gets a reference to the value exactly located at the path indicated.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
+        self.root.get(path)
+    }
+}
+
+impl<V: Encode> Encode for BinaryTree<V> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.root.encode(bytes)
+    }
+}
+
+impl<V: Decode> Decode for BinaryTree<V> {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            root: Node::decode(bytes)?,
+        })
+    }
+}
+
+impl<V: Display> Display for BinaryTree<V> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        writeln!(f, "■\n{}", self.root)
+    }
+}
+
+/// Represents a node of a binary tree.
+#[derive(Debug)]
+struct Node<V> {
+    value: V,
+    left: Option<Box<Node<V>>>,
+    right: Option<Box<Node<V>>>,
+}
+
+impl<V> Node<V> {
+    fn new(value: V) -> Self {
+        Self {
+            value,
+            left: None,
+            right: None,
+        }
+    }
+
+    fn insert(&mut self, path: &Path, value: V) {
+        let new_node = Some(Box::new(Node::new(value)));
+        let mut iter = Some(self);
+
+        for bit in path.iter() {
+            let node = iter.take().unwrap();
+
+            if !bit {
+                match node.left {
+                    None => {
+                        node.left = new_node;
+                        break;
+                    }
+                    Some(_) => iter = node.left.as_deref_mut(),
+                };
```

Comment:
> We could either return an error or panic if we find an empty left/right pointer too early, or make node.value an option, and keep building the rest of the path's nodes before trying to store the value (this may affect the serialization format).

I opted by the former.